pyo3-build-config/src/impl_.rs

```diff
@@ -177,12 +177,19 @@ impl InterpreterConfig {
             PythonImplementation::GraalPy => out.push("cargo:rustc-cfg=GraalPy".to_owned()),
         }
 
-        if self.abi3 {
+        // If Py_GIL_DISABLED is set, do not build with limited API support
+        if self.abi3 && !self.build_flags.0.contains(&BuildFlag::Py_GIL_DISABLED) {
             out.push("cargo:rustc-cfg=Py_LIMITED_API".to_owned());
         }
 
+        if self.build_flags.0.contains(&BuildFlag::Py_GIL_DISABLED) {
+            out.push("cargo:rustc-cfg=Py_GIL_DISABLED".to_owned());
+        }
+
         for flag in &self.build_flags.0 {
-            out.push(format!("cargo:rustc-cfg=py_sys_config=\"{}\"", flag));
+            if flag != &BuildFlag::Py_GIL_DISABLED {
+                out.push(format!("cargo:rustc-cfg=py_sys_config=\"{}\"", flag));
+            }
```

Comment:
Might be a little tidier to use a `match` here, and then can drop the if just above:

```suggestion
            match flag {
                BuildFlag::Py_GIL_DISABLED => out.push("cargo:rustc-cfg=Py_GIL_DISABLED".to_owned());
                flag => out.push(format!("cargo:rustc-cfg=py_sys_config=\"{}\"", flag));
            }
```